### ISSUE TITLE
Allow data export from the Grid

### ISSUE THREAD

**Issue by gsobolev** [DevExpress/devextreme-reactive#537]
Provide the capability to export data in different formats (CSV, Excel, etc.)

**Comment by gsobolev** [comment]
While the Grid does not have any built-in export capabilities, there is a way to work around this limitation. It is possible to obtain data that is displayed by the table and then export it manually in the required format.
Please check [this example](https://stackblitz.com/edit/react-gkzc22?file=index.js) to see how the display data can be obtained.

**Comment by mannylau** [comment]
The example link is down. Please advise.

**Comment by gsobolev** [comment]
Hi @mannylau,

I have checked the link and it works fine for me. Are you still having this problem?

**Comment by gsobolev** [comment]
@samuelcotterell, 
 
The '1.0.0-beta.1' version of React Grid is used in the example provided in this thread. We have made some API changes since that beta. I have created a separate example with the latest version. Please find it [here](https://stackblitz.com/edit/react-2xtzk8?file=index.js).

**Comment by samuelcotterell** [comment]
great! thanks for that.

**Comment by angaaron** [comment]
Hi @gsobolev thanks for that example. I was wondering how this would be done with a table with multiple pages? Seems to only extra data from the first page. Cheers

**Comment by gsobolev** [comment]
Hi @angaaron!

`TemplateConnector` allows you to access already processed data. So you need to use a bit different approach in the described case. Please try to use [Getter](https://devexpress.github.io/devextreme-reactive/react/core/docs/reference/getter/) like in the [following demo](https://stackblitz.com/edit/react-3rvkcf?file=index.js).

Note that exported data will vary depending on the 'Getter' position.

**Comment by angaaron** [comment]
This is perfect! Thanks heaps @gsobolev !

**Comment by susuwatari2** [comment]
Hi guys, can explain what export code is? @gsobolev

**Comment by ppaul** [comment]
I was very happy to see this functioning. However, after more customizations I found out that it ignores column visibility (hiding some columns) and column ordering. If I put the Getter after those plugins, it gets no rows. Is there anything that should be done in certain way? Thanks

**Comment by MaximKudriavtsev** [comment]
Hi @ppaul,

The order of integrated and custom plugins can change your Grid's data processing logic. See this [guide](https://devexpress.github.io/devextreme-reactive/react/grid/docs/guides/plugin-overview/) for more information.

**Comment by MaximKudriavtsev** [comment]
@ppaul 

To get the required behavior, we recommend you create a custom plugin. All plugins are based on our React Core components. Follow this [guide](https://devexpress.github.io/devextreme-reactive/react/core/docs/guides/fundamentals/) for more information.
 
Create a custom plugin `MyExportPlugin` that allows access to Grid `getters`. I used `rows` and `tableColums` getters to create the plugin. You can see what `getters` are available in our documentation in the `Plugin Developer Reference` section. For example, the `TableColumnReordering` plugin exports two getters - `tableColumns` and `tableHeaderRows`. Refer to [link](https://devexpress.github.io/devextreme-reactive/react/grid/docs/reference/table-column-reordering/#plugin-developer-reference) for more information.
 
See this [sample](https://codesandbox.io/s/x9325zo10q) with the custom MyExportPlugin.

**Comment by ppaul** [comment]
@MaximKudriavtsev , I come back with a new question. It was clear with the previous ones. But now that I implemented as in the sample, I notice that the <DataTypeProvider /> we have, is not considered. As of order, I specified the plugin as last, so it should have taken into account the provider. Am I missing something here? The data type provider specifies some transformations that are applied on the cells with formatterComponent and for. What I want to achieve is having the same data collection as is displayed in the table. Thank you for your advices so far!

**Comment by ppaul** [comment]
Hello @MaximKudriavtsev ,
Unfortunately I cannot post the code due to privacy reasons. However, what I am saying is this: if custom formatters are used, with getCellValue functions attached, these functions come into tableColumns, but I have to call them (again) manually. It would be nice if the data would be available as it is seen, with these functions already applied. We have a language dependent app, therefore with the same data, the content of the cells vary. In summary, the set-up is like this:
```
 <Grid rows={dataSource} columns={columns}}> //'@devexpress/dx-react-grid-material-ui'
          // sorting component
          // dataTypeProvider component, has getCellValue functions defined
          <VirtualTable  columnExtensions={tableColumnExtensions} cellComponent={VirtualTableCell}        
                  // row component
              />
              <TableColumnReordering order={columnOrder} onOrderChange={changeColumnOrder} />
              <ExcelPlugin /> // the problem is that the data is still raw here, no formatters applied
       </Grid>
```

**Comment by MaximKudriavtsev** [comment]
The way you are suggesting is the best solution now - use the `getCellValue` function in the `ExcelPlugin`.
 
However, Data Export is a feature that we will consider to implement in the future. We would greatly appreciate your feedback.

**Comment by ciforusalterra** [comment]
Hi, any update here about exporting table to excel or even pdf? I need formated data to be exported, not just raw data. Thanks.

**Comment by SergeyAlexeev** [comment]
[We have plans](https://community.devexpress.com/blogs/javascript/archive/2019/01/22/devextreme-javascript-2019-roadmap.aspx) to implement it in the second half of 2019. Please take a look at the 'Native React Grid' section.

**Comment by MaximKudriavtsev** [comment]
Hi all,

We started to implement this feature. Please, follow this [topic](https://github.com/DevExpress/devextreme-reactive/issues/2366) to be informed about our updates. We would also appreciate your feedback on this feature. You can provide it in the same topic.

**Comment by serche** [comment]
Hi guys, any plan to deliver excel export in near future?

**Comment by ushkal** [comment]
Hi @serche ,
This feature has high priority and we plan to implement it soon.

**Comment by flyfishMT** [comment]
Hello, I am trying to export in a Plugin.  My Plugin is before IntegratedPaging in the Grid, but I still only get a page of data.  [Here is a sample](https://stackblitz.com/edit/react-c5fctj).

**Comment by AryamnovEugeniy** [comment]
Hello @flyfishMT ,

Even when you create a custom plugin, you need to use Getter in this plugin, as you've done in `index.js`,  to get the required behaviour. I've modified your [example](https://stackblitz.com/edit/react-t1aoq3?file=MyPlugin.js) and use Getter instead of TemplateConnector.
To the point, our export feature will be implemented very soon and you will be able to use it instead (you can follow [this thread](https://github.com/DevExpress/devextreme-reactive/pull/2631) to be informed about our updates).

**Comment by MaximKudriavtsev** [comment]
We have released the `2.5.0` version of [DevExtreme Reactive Grid](https://devexpress.github.io/devextreme-reactive/react/grid/docs/guides/export/) with Excel export feature.
